tests/MongoDB.Bson.Tests/IO/JsonReaderTests.cs

```diff
@@ -116,18 +121,114 @@ public void TestArrayTwoElements()
         [InlineData("{ $binary : \"AQ==\", $type : \"00\" }", new byte[] { 1 }, BsonBinarySubType.Binary)]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"0\" } }", new byte[] { 1 }, BsonBinarySubType.Binary)]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"00\" } }", new byte[] { 1 }, BsonBinarySubType.Binary)]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\" } }", new byte[] { 1 }, BsonBinarySubType.Binary)]
+        [InlineData("{ $binary : { subType : \"00\", base64 : \"AQ==\" } }", new byte[] { 1 }, BsonBinarySubType.Binary)]
         [InlineData("BinData(0, \"AQ==\")", new byte[] { 1 }, BsonBinarySubType.Binary)]
+        [InlineData("HexData(0, \"01\")", new byte[] { 1 }, BsonBinarySubType.Binary)]
         public void TestBinaryData(string json, byte[] expectedBytes, BsonBinarySubType expectedSubType)
         {
             using (var reader = new JsonReader(json))
             {
                 var result = reader.ReadBinaryData();
 
                 result.Should().Be(new BsonBinaryData(expectedBytes, expectedSubType));
+                reader.State.Should().Be(BsonReaderState.Initial);
                 reader.IsAtEndOfFile().Should().BeTrue();
             }
         }
 
+        [Theory]
+        [InlineData("{ $binary")]
+        [InlineData("{ $binary :")]
+        [InlineData("{ $binary : \"AQ==\"")]
+        [InlineData("{ $binary : \"AQ==\",")]
+        [InlineData("{ $binary : \"AQ==\", $type")]
+        [InlineData("{ $binary : \"AQ==\", $type :")]
+        [InlineData("{ $binary : \"AQ==\", $type : 0")]
+        [InlineData("{ $binary : {")]
+        [InlineData("{ $binary : { base64")]
+        [InlineData("{ $binary : { base64 :")]
+        [InlineData("{ $binary : { base64 : \"AQ==\"")]
+        [InlineData("{ $binary : { base64 : \"AQ==\",")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType :")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"0\"")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"0\" }")]
+        [InlineData("{ $binary : { subType")]
+        [InlineData("{ $binary : { subType :")]
+        [InlineData("{ $binary : { subType : \"0\"")]
+        [InlineData("{ $binary : { subType : \"0\",")]
+        [InlineData("{ $binary : { subType : \"0\", base64")]
+        [InlineData("{ $binary : { subType : \"0\", base64 :")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\"")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\" }")]
+        [InlineData("BinData")]
+        [InlineData("BinData(")]
+        [InlineData("BinData(0")]
+        [InlineData("BinData(0,")]
+        [InlineData("BinData(0, \"AQ==\"")]
+        [InlineData("HexData")]
+        [InlineData("HexData(")]
+        [InlineData("HexData(0")]
+        [InlineData("HexData(0,")]
+        [InlineData("HexData(0, \"01\"")]
```

Comment:
Test every single place where the binary extended JSON is valid right up to the point where end of file is reached.